flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/FunctionQueryOperation.java

```diff
@@ -0,0 +1,116 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.operations;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.catalog.ContextResolvedFunction;
+import org.apache.flink.table.catalog.ResolvedSchema;
+import org.apache.flink.table.expressions.ResolvedExpression;
+import org.apache.flink.table.expressions.TableReferenceExpression;
+import org.apache.flink.table.types.DataType;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/** Describes a relational operation that was created from applying a (process) table function. */
+@Internal
+public class FunctionQueryOperation implements QueryOperation {
+
+    private static final String INPUT_ALIAS = "$$T_FUNC";
+
+    private final ContextResolvedFunction resolvedFunction;
+    private final List<ResolvedExpression> arguments;
+    private final ResolvedSchema resolvedSchema;
+
+    public FunctionQueryOperation(
+            ContextResolvedFunction resolvedFunction,
+            List<ResolvedExpression> arguments,
+            ResolvedSchema resolvedSchema) {
+        this.resolvedFunction = resolvedFunction;
+        this.arguments = arguments;
+        this.resolvedSchema = resolvedSchema;
+    }
+
+    public ContextResolvedFunction getResolvedFunction() {
+        return resolvedFunction;
+    }
+
+    public List<ResolvedExpression> getArguments() {
+        return arguments;
+    }
+
+    public DataType getOutputDataType() {
+        // Make sure time attributes are not erased
+        final List<String> fieldNames = resolvedSchema.getColumnNames();
+        final List<DataType> fieldTypes = resolvedSchema.getColumnDataTypes();
+        return DataTypes.ROW(
+                        IntStream.range(0, fieldNames.size())
+                                .mapToObj(
+                                        pos ->
+                                                DataTypes.FIELD(
+                                                        fieldNames.get(pos), fieldTypes.get(pos)))
+                                .collect(Collectors.toList()))
+                .notNull();
```

Comment:
Could we extract that to `ResolvedSchema` or at least a util class? We do that at least twice.

Reply:
This logic is very fragile. I will add a utility method but we should rework the way how we deal with time columns in Table API. Once we switch to SQL based serialization, we don't need it anymore.